### ISSUE TITLE
Transclude deeply?

### ISSUE THREAD

**Issue by petejohanson** [inadarei/draft-prefer-transclude#5]
For media types in responses that allow deep nesting, .e.g. HAL embedded resources, it would be nice to support deeply transcluding a series of link relations.

For example, when fetching a blog post, we want to transclude the comments, and the profiles of the users posting those comments:

    Prefer: transclude="comments.user-profile"

Or some such. I'm not sure the syntax that would be most natural, but allowing this feature combined with tailored profiles/media types would allow some if the benefits of GraphQL in combining/reducing requests.

Thoughts?

**Comment by asbjornu** [comment]
How does this relate to the [OData Query Options `$expand` keyword](http://docs.oasis-open.org/odata/odata/v4.0/errata03/os/complete/part2-url-conventions/odata-v4.0-errata03-os-part2-url-conventions-complete.html#_Toc453752359)?

**Comment by petejohanson** [comment]
Very similar in spirit. OData's Navigation Properties are akin to link relations in practice for this, and $expand does allow specifying a sequence of Properties, e.g. Products/Description.

I see this RFC as standing on the shoulders of existing hypermedia concepts, e.g. link relations, instead of reinventing everything, a la OData.

**Comment by asbjornu** [comment]
I've proposed to split the OData Query Options out into its own specification (OData/odataorg.github.io#111). I thought that there might be something to be learned from how OData's query keywords are defined, perhaps not stopping at transclusion, but looking at OData's full feature set and attempt to reach feature parity.

It would be nice to have open standard alternatives to proprietary technologies such as [GraphQL](http://graphql.org/) and [Falcor](https://netflix.github.io/falcor/).

**Comment by asbjornu** [comment]
Perhaps the [JSON Pointer (RFC 6901)](https://tools.ietf.org/html/rfc6901) syntax can be used to express which nodes to transclude deeply?

**Comment by inadarei** [comment]
@petejohanson deep transclusion is very interesting. I like the dot-syntax for sure. Do you still think it's the best way to go or has your view changed on it?

Thank you!

**Comment by petejohanson** [comment]
As nice as the dot-syntax is, I think it would be insufficient for some media types, scenarios.

For example, link relations in HAL can *either* be IANAL standard link relations, e.g. `next` or `collection`, complete URIs, e.g. `https://api.mycompany.com/rels/order`, or CURIEs e.g. `myapp:order`.

Using `.` as the separator would then require escaping the `.`s found in the individual link relations, or some other awkward handling.

Brainstorming on this, I began to wonder if something inspired by bash expansion (which has the advantage of using a character that usually gets percent encoded and isn't present in URIs/CURIEs), like:

```
Prefer: transclude="next{myapp:user},previous{myapp:user}"
```

Which would transclude the next -> myapp:user and previous -> myapp:user

Or:

```
Prefer: transclude="next{myapp:user,myapp:category}"
```

Which would include the next item, and transclude the myapp:user and myapp:category of that next item.

When transcluding very deeply, however, this does result in awkward {} nesting, e.g.:

```
Prefer: transclude="next{myapp:user{myapp:address},myapp:category{myapp:recent{item}}}"
```

Pros:

* Allows a single transclude directive to nest a "graph" of resources, not just a linear line.
* Allows using URIs as transclusion targets, without awkward escaping requirements.
* Borrows from a precedent, with bash expansion as the origin.

Cons:

* Requires matching open and close brackets, complicating the simple linear transclusion case.

## Alternative Proposal

The other idea I had, was the same idea to use reserved characters, and still avoid the `.` for chaining.

Simple scenario using standard link relations:

```
Prefer: transclude="next ! collection ! profile"
```

Simple scenario using CURIEs:

```
Prefer: transclude="next ! myapp:user ! myapp:post"
```

Transcluding several link relations deeper in the 'chain':

```
Prefer: transclude="next ! [myapp:user,myapp:category]"
```

Basically, use `!` to chain linearally, and if you want to transclude several relations at a given level, group with `[` and `]`. Additionally, you can optionally use spaces before/after `!` and `,` to make it more readable.

Pros

* The simple linear transclusion chain is succinct, and fairly readable.
* Avoids use of `.` in syntax, to avoid awkward escaping for URIs link relations.

Cons

* `!` character as linear chain is not as familiar as `.` character.
* Increases the size of the header slightly to encode the same information.

Thoughts on either?

**Comment by asbjornu** [comment]
@petejohanson: Since space is usually a disallowed character in tokens such as link relations, perhaps space can be the chain separator and then use comma as a way to transclude different branches? Sort of like CSS:

```http
Prefer: transclude="next collection profile, next myapp:user myapp:post"
```

**Comment by inadarei** [comment]
@asbjornu's suggestion sounds very elegant. 

Couple notes: 
1. The separator for multiple transclusions in the current transclude spec is ";" not a comma, I assume that is fine.
2. What we are talking here about really is "pointer to JSON element under which link rel is". Because link rels themselves are not hierarchical, not in the initial example of HAL given. 

Given #2, I think what makes sense is to indicate parent JSON element with JSON Pointer (RFC 6901) as @asbjornu suggested and then indicate the actual name of the link rel after space (so leveraging both of @asbjornu's recommendations). This way it will be very clean and clear which part is path and which part is the link rel?

For instance, if we take the example HAL from HAL's website:

```
{
    "_links": {
        "self": { "href": "/orders" },
        "curies": [{ "name": "ea", "href": "http://example.com/docs/rels/{rel}", "templated": true }],
        "next": { "href": "/orders?page=2" },
        "ea:find": {
            "href": "/orders{?id}",
            "templated": true
        },
        "ea:admin": [{
            "href": "/admins/2",
            "title": "Fred"
        }, {
            "href": "/admins/5",
            "title": "Kate"
        }]
    },
    "currentlyProcessing": 14,
    "shippedToday": 20,
    "_embedded": {
        "ea:order": [{
            "_links": {
                "self": { "href": "/orders/123" },
                "ea:basket": { "href": "/baskets/98712" },
                "ea:customer": { "href": "/customers/7809" }
            },
            "total": 30.00,
            "currency": "USD",
            "status": "shipped"
        }, {
            "_links": {
                "self": { "href": "/orders/124" },
                "ea:basket": { "href": "/baskets/97213" },
                "ea:customer": { "href": "/customers/12369" }
            },
            "total": 20.00,
            "currency": "USD",
            "status": "processing"
        }]
    }
}
```

There are two `ea:basket` link rels, if we want to ask to only transclude teh first one we would request:

```
Prefer: transclude="/_embedded/ea:order/0 ea:basket"
```

If we want to indicate both, explicitely:

```
Prefer: transclude="/_embedded/ea:order/0 ea:basket;/_embedded/ea:order/1 ea:basket"
```

And here's where JSON Pointer's elegance can be used, we can also indicate all "ea:basket"s with a syntax like:

```
Prefer: transclude="/_embedded/ea:order ea:basket"
```

because in JSON pointers `/_embedded/ea:order` actually means: "both elements"

Makes sense?

**Comment by petejohanson** [comment]
@inadarei That works fine for folks using 'HAL CURIEs", but for anyone using full URLs for custom link relations, that could get messy... Per https://tools.ietf.org/html/rfc6901#section-3 you'd end up w/ something like:

```
Prefer: transclude="/_embedded/https:~1~1example.com~1rels~1order http:~1~1example.com~1rels~1basket"
```

I wonder what percentage of APIs in the wild actually conform to the link rel spec, and don't use CURIEs, so they have full URIs for non-standard/registered link relations. Perhaps it's not enough to worry?

**Comment by inadarei** [comment]
@petejohanson good point on URIs but I believe in non-HAL what you would get is more something like this:

```
Prefer: transclude="/something/order http://example.com/rel/basket"
```

URI shouldn't show up in the JSON pointer, because nobody really names random JSON properties with URIs, it may show up in the link rel itself, correct?

**Comment by ghost** [comment]
I suppose this may be obvious, but using JSON Pointer for this would really only work for JSON-based media types, and the transclude spec is supposed to be agnostic of media type, isn't it?

Also, regarding the use of the `.` notation for separating links in the chain, is it really that big of a deal to escape the `.` that could be found in link relations?  I can't imagine dealing with that is really hard for a server to handle, or hard for a client to construct.

**Comment by asbjornu** [comment]
I agree with @scottsmith130 and withdraw my suggestion to use JSON Pointer. We need a media type agnostic query language. For that I think a subset of CSS Selectors could work quite well. 

The alternative would be to indicate the media type of the query language itself in the header with a parameter, or something like that. But how would that work with several different query languages and so forth? Sounds messy.

**Comment by inadarei** [comment]
Well, technically JSON PATH is for JSON only, but - with the same token, you could argue that CSS is only defined for HTML?

I think the *syntax* of JSON PATH (not the actual RFC) can be defined for any hierarchical markup language, including XML, no?

**Comment by ghost** [comment]
Well, sure JSON Pointer syntax could be defined to apply to other formats. But it only defines how it applies to JSON so it would be up to some other spec to define how it applies to another format. Assuming that is reasonable then sure JSON Pointer-like syntax could still be used here.

**Comment by asbjornu** [comment]
CSS already works with [XML](https://www.w3.org/Style/styling-XML.en.html) and [JSON](https://github.com/lloyd/JSONSelect). The CSS Selector language is also very expressive and battle tested, so it feels like a safe choice. If higher levels than a subset of CSS Selector Level 1 is warranted, no work needs to be done; Level 2, 3 and 4 already exists.

**Comment by inadarei** [comment]
It feels like we have a solution that most people on this thread agree to and that looks elegant + leverages existing standards. This issue has also been open long-enough that anybody who may have wanted to voice an opinion, probably had an opportunity. I think we should go ahead and incorporate the consensus solution, which is CSS.

Is anybody up to summarize the final solution in a PR? I am thinking a new section about Deep Transclusion, with explanation, solution and examples?

Thank you,

**Comment by petejohanson** [comment]
The CSS selector approach seems ok, I'm curious how we would apply that to less straight forward transclusion "chains". I think we'd need to use CSS Selector Level 2 at a minimum to make this worth-while:

```
next collection:first-child post
```

If we got that route, we'd need to clearly define how various pseudo selectors, etc. are expected to behave.

**Comment by zdne** [comment]
@petejohanson @asbjornu would you please provide a more complex example of CSS-like syntax? Thanks!

**Comment by asbjornu** [comment]
@zdne, sure! If we take @inadarei's JSON from above:

```json
{
    "_links": {
        "self": { "href": "/orders" },
        "curies": [{ "name": "ea", "href": "http://example.com/docs/rels/{rel}", "templated": true }],
        "next": { "href": "/orders?page=2" },
        "ea:find": {
            "href": "/orders{?id}",
            "templated": true
        },
        "ea:admin": [{
            "href": "/admins/2",
            "title": "Fred"
        }, {
            "href": "/admins/5",
            "title": "Kate"
        }]
    },
    "currentlyProcessing": 14,
    "shippedToday": 20,
    "_embedded": {
        "ea:order": [{
            "_links": {
                "self": { "href": "/orders/123" },
                "ea:basket": { "href": "/baskets/98712" },
                "ea:customer": { "href": "/customers/7809" }
            },
            "total": 30.00,
            "currency": "USD",
            "status": "shipped"
        }, {
            "_links": {
                "self": { "href": "/orders/124" },
                "ea:basket": { "href": "/baskets/97213" },
                "ea:customer": { "href": "/customers/12369" }
            },
            "total": 20.00,
            "currency": "USD",
            "status": "processing"
        }]
    }
}
```

The following queries can be used to select various properties:


```css
 /* Select 'self', 'curies', 'next', 'ea:find', 'ea:admin', 'self', 'ea:basket' and 'ea:customer' */
_links > *

/* Selects the first 'self' link's 'href' property within '_embedded' */
_embedded _links self:first-of-type href

/* Selects the 'href' of 'ea:customer' inside the 'ea:order' object that has 'status' set to something not equal to 'processing' */
ea:order *:not([status="processing"]) + _links ea:customer href
```

The [JSONSelectTests repository](https://github.com/lloyd/JSONSelectTests) contains a lot of examples of the kinds of queries it's possible to perform.

**Comment by dret** [comment]
reading through this now i am not 100% sure this is a good 80/20 deal. i am not doubting the fact that something *could* be done. but please try to balance the complexity (this should not be tightly coupled to any media type) and the gains. unless you're ok with this ballooning into some rather complex meta-graphql language framework, i think this maybe one of those things that could be done, but for *implementation complexity vs. gains* reasons maybe shouldn't be done.

**Comment by gabesullice** [comment]
Can we simply _not_ have this specification dictate how one selects a subset of links to be transluded (e.g. by using JSON Pointer or CSS selectors)? Instead, could we reuse the URI fragment concept?

[RFC3986](https://tools.ietf.org/html/rfc3986#section-3.5) has this:

> Fragment identifiers have a special role in information retrieval systems as the primary form of client-side indirect referencing, allowing an author to specifically identify aspects of an existing resource that are only indirectly provided by the resource owner.

...

> The fragment's format and resolution is therefore dependent on the media type [RFC2046] of a potentially retrieved representation

...

> Individual media types may define their own restrictions on or structures within the fragment identifier syntax for specifying different types of subsets

That would allow for `application/vnd.api+json` to define JSON Pointer as the fragment syntax for JSON:API and for `application/hal+xml` to define XPath as the fragment synta. This spec could then remain completely agnostic about a link selection syntax.

Meaning:

`Prefer: transclude="#article-0 author"` would work for HTML, but
`Prefer: transclude="#/data/0/relationships/author self"` would work for JSON:API.

**Comment by asbjornu** [comment]
@gabesullice: That would not allow us to transclude more than one level, but I suppose from our discussion on Twitter that one level might be enough for now. This at least gets the ball moving and might be a good compromise. The most important aspect is that it's content-type agnostic, which is a requirement. Thumbs up from me! 👍

**Comment by evert** [comment]
I don't think HAL typically needs anchors though. Afaik there's only ever the top level links.

**Comment by petejohanson** [comment]
In principle, the fragment approach seems to align nicely, but I'm not sure how nicely that works when media types use [extension link relations](https://tools.ietf.org/html/rfc5988#section-4.2) which are full URIs.... Now you've got to somehow stick a URI inside a fragment? My brain hard freezes when considering the encoding/decoding implications of such a thing.

**Comment by petejohanson** [comment]
@evert True, for a single resource, but when you consider that you may have embedded resources (which can be used for other uses beyond "transclusion"), you may still need to deeply reference what you want.

**Comment by evert** [comment]
Embedded resources are already 'transcluded'. They have their own urls. The sane way to traverse those links too would be with multi-level traversal, but ideally _embedded goes away entirely, replaced by H/2 push

**Comment by gabesullice** [comment]
>  That would not allow us to transclude more than one level, but I suppose from our discussion on Twitter that one level might be enough for now.

@asbjornu I don't see how using fragments precludes depth-N. Piggybacking off @evert's ideas, one could easily use parenthesis to indicate nesting in a LISPy style:

```
Prefer: transclude="(#/data/0/relationships/author self (#/data/links/profile-photo enclosure)); ..."
```

or, using a pipe:

```
Prefer: transclude="#/data/0/relationships/author self | #/data/links/profile-photo enclosure; ..."
```

> Now you've got to somehow stick a URI inside a fragment? My brain hard freezes when considering the encoding/decoding implications of such a thing.

@petejohanson, I think you might have misunderstood the syntax I proposed. An extension link relation would not need to appear in a fragment. (It was probably my fault for using a fragment that had `author` in it, that was an unfortunate coincidence).

Perhaps this will clear it up:

```
Prefer: transclude="#/data/0/relationships/writer self"
                    \___________________________/ \__/
                                   |                |
                                fragment      link-relation
```

The fragment identifies the _context_ of the link, not the link itself. The document I was imagining looked like this:

```
{
  "data": [{
    "type": "article",
    "id": 1,
    "relationships": {
      "writer": {
        "links": {
          "self": "/users/1"
        }
      }
    }
  }, {
    ...
  }]
}
```
